src/hooks/useQueryState.js

```diff
@@ -0,0 +1,163 @@
+      // note: must include toString in defaultVal.toString() in order to check empty arrays
+      if (
+        (urlParams.get(key) ?? defaultVal.toString()) === state.toString() ||
+        (state === true && urlParams.get(key) === '')
+      )
+        return;
+
+      // key with no value OR boolean key set to false -> absent from url, push empty
+      if (!state || state === '' || state.length === 0 || state === false) urlParams.delete(key);
+      // boolean or value key is present in url with its current value
+      else urlParams.set(key, state.toString());
+
+      // boolean keys set to true don't need '=value'
+      urlParams = urlParams
+        .toString()
+        .replace(/=true|=$/g, '')
+        .replace(/=&/, '&');
+
+      history.push('?' + urlParams);
+    };
+    setURLParam();
+    /*********************** ESLint Disable Justification: *******************************/
+    // ESLint wants to include `defaultVal, key, history` as dependencies but to prevent
+    // extra loops of ~ url setting state setting the url ~ we do not include.
+    // The only time we want to set the URL manually is when the state changes
+    /*************************************************************************************/
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [state]);
+
+  /**
+   * Anytime the url is changed, loadURLParams is invoked
+   * This function pulls the value of the key from the url
+   *   and updates the state with it (if changed)
+   *
+   * This is a useLayoutEffect rather than useEffect in order to prioritize loadURLParams;
+   *   this is necessary for preventing the race condition where (for ex.) the user clicks 'back'
+   *   twice quickly and the first 'back' sets the URL params before the second 'back' can
+   *   load the accurate new URL param. This is still not totally perfect but only breaks when
+   *   a user really messes with the navigation significantly and quickly.
+   * Comparison of the two effects: https://kentcdodds.com/blog/useeffect-vs-uselayouteffect
+   */
+  useLayoutEffect(() => {
+    const loadURLParams = async () => {
+      const urlParams = new URLSearchParams(history.location.search);
+      let urlValue = urlParams.get(key);
+
+      switch (type) {
+        // using boolean key -> load true if present, false if absent from url
+        case types.Boolean:
+          setState(urlParams.has(key));
+          return;
+
+        // different logic if we are working with an array than with single value
+        // because setState will not do deep array comparison
+        case types.Array:
+          // if the same value (or unset and state array is empty) -> do nothing
+          // this check is necessary because [] -> [] are different memory objects
+          if (urlValue === state.toString() || (!urlValue && state.length === 0)) return;
+          else if (!urlValue) setState(defaultVal);
+          else setState(urlValue.split(','));
```

Comment:
Would it be clearer to combine these?:
```suggestion
          else setState(urlValue ? urlValue.split(',') : defaultVal);
```
Or since `urlValue` is `null` when `key` is not present in the url, you could even do:
```suggestion
          else setState(urlValue?.split(',') ?? defaultVal);
```